### PR TITLE
JSON AST: replace ``isConstructor`` by ``kind`` which also supports fallbacks

### LINKED ISSUE #4691: JSON AST: remove ``is`` prefix, check for ``fallback``
Came up in https://github.com/ethereum/solidity/pull/4663#discussion_r207832105

``isConstructor`` in the JSON AST should be changed to ``constructor``.
We should verify that fallback functions are in the AST.

### PR DESCRIPTION
**Pull request by ekpyron**
Fixes #4691.

### CODE REVIEW

File: libsolidity/ast/ASTJsonConverter.cpp [hunk 1]
```diff
@@ -336,6 +335,14 @@ bool ASTJsonConverter::visit(FunctionDefinition const& _node)
 		make_pair("implemented", _node.isImplemented()),
 		make_pair("scope", idOrNull(_node.scope()))
 	};
+	if (m_legacy)
```
Comment (axic) [inline libsolidity/ast/ASTJsonConverter.cpp:339]: I think there's no harm adding `kind` to legacy mode as long as `isConstructor` is kept.

Comment (ekpyron) [reply libsolidity/ast/ASTJsonConverter.cpp:339]: I adjusted the last commit accordingly.

**Comment by axic** [comment]
@ekpyron I was wondering if it should be named `functionKind` similarly how `ContractDefinition` has `contractKind`, then that reflects the internal field we have. We do not have such a field for the `FunctionDefinition` AST node, but I think we should.

**Comment by ekpyron** [comment]
@axic Yes, I was wondering the same. ``FunctionCall`` has a ``kind`` and ``Literal`` has a ``kind``, though. Not sure whether it's better to change ``contractKind`` to ``kind`` or whether to change the one in this PR to ``functionKind`` (and if the latter whether to change the other ``kind``s as well).

**Comment by axic** [comment]
If `kind` instead of `xyzKind` is more common, then lets stick to `kind`.

**Comment by ekpyron** [comment]
Well - it's two against one, so *slightly* more common. I tend to stick to ``kind``.

**Comment by axic** [comment]
As a second look we have:
- `enum ContractDefinition::ContractKind` and `ContractDefinition::contractKind()` -> AST: `contractKind`
- `enum FunctionCallKind` and `FunctionCallAnnotation::kind` -> AST: `kind`
- `enum ArrayType::ArrayKind` and `ArrayType::arrayKind` -> AST: (not present)
- `enum FunctionType::FunctionKind` and `FunctionType::kind()`

And the `kind` field for `Literal` is specific to the AST JSON, derived from the token type by `ASTJsonConverter::literalTokenKind`.

**Comment by ekpyron** [comment]
The ``FunctionType::kind()`` is more low-level, though, and wouldn't be the same as the new ``kind`` we want to introduce here.
Maybe we should discuss the whole JSON AST in a separate issue - not sure whether we should merge PR first or postpone it, resp. close it for now, though.

**Comment by chriseth** [comment]
Moving back to "in progress" as it seems that this still requires discussion.

**Comment by ekpyron** [comment]
Moving back to "under review", since the PR is still up to date and we decided to probably just merge it.

**Comment by axic** [comment]
We've agreed on a daily call to go with `kind` and perhaps rename `contractKind` in the future.

**Pull request merged**